plugins/inputs/x509_cert/x509_cert.go

```diff
@@ -246,6 +250,16 @@ func (c *X509Cert) collectCertURLs() ([]*url.URL, error) {
 // Gather adds metrics into the accumulator.
 func (c *X509Cert) Gather(acc telegraf.Accumulator) error {
 	now := time.Now()
+
+	if c.tlsCfg.ServerName != "" && c.ServerName == "" {
+		// Save SNI from c.tlsCfg.ServerName to c.ServerName and reset c.tlsCfg.ServerName.
+		// We need to reset c.tlsCfg.ServerName for each certificate when there's
+		// no explicit SNI (c.tlsCfg.ServerName or c.ServerName) otherwise we'll always use
+		// first uri HostName for all certs (see issue 8914)
+		c.ServerName = c.tlsCfg.ServerName
+		c.tlsCfg.ServerName = ""
+	}
```

Comment:
I'm not sure I understand why you need to do this. Why not just reset `c.tlsCfg.ServerName`?

Reply:
Purpose for this is that if you've a config like:
```
[[inputs.x509_cert]]
  sources = [ "https://www.example.com:443", "tcp://example.org:443", "https://some.domain.com:443" ]
  tls_server_name = "why.example.com"
  # or
  #server_name = "why.example.com"
```
Then you can "force" all certs/requests to have why.example.com SNI. I'm not sure if this is a real use case, but this way you can force SNI (and use that SNI for validation) ?
I think w/out this first request will have why.example.com SNI and other requests will use example.org and some.domain.com (because c.tlsCfg.ServerName is reset between requests).

Reply:
Yeah but shouldn't it be `c.ClientConfig.ServerName` then? As far as I can see, this `if` condition is only ever true the first time `Gather()` is called as afterwards `c.tlsCfg.ServerName` will _always_ be empty. If this is the intention, then please move this code to `Init()` and we are on the same page. :-)